### PR TITLE
Add edit all children to the scene dock

### LINKED ISSUE #856: Add edit all children button
**Describe the project you are working on:**

3d game project.

**Describe the problem or limitation you are having in your project:**

Make instanced scene and all children editable because it is a quality of life problem with many nested scenes.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

Add a button called "make all children editable".

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

The button will be added to the right click menu. This button will go through the nested childs of the scene and make that and the instanced scene editable.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

This is an editor feature and is expected to be accessible to new developers.

**Is there a reason why this should be core and not an add-on in the asset library?:**

Downloading the asset from the library means you'll need to explain the asset store and that is not common for new developers.

### PR DESCRIPTION
**Pull request by fire**
Sponsored by IMVU.

Fixes https://github.com/godotengine/godot-proposals/issues/856

Edit:

video incoming.

GIF 2020-06-04 4-25-09 PM

### CODE REVIEW

**Comment by fire** [comment]
Test project.
[Test Project Project.zip](https://github.com/godotengine/godot/files/4733218/Test.Project.Project.zip)

File: editor/scene_tree_dock.cpp [hunk 1]
```diff
@@ -1749,17 +1763,26 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
...
-		if (editable) {
-			p_node->set_scene_instance_load_placeholder(false);
-		}
+void SceneTreeDock::_toggle_editable_children(Node *p_node, Node *p_owner, bool p_edit_all) {
+	if (!p_node) {
+		return;
+	}
+	bool editable = !EditorNode::get_singleton()->get_edited_scene()->is_editable_instance(p_node);
+	bool root_editable = EditorNode::get_singleton()->get_edited_scene()->is_editable_instance(p_owner);
+	editable |= !root_editable;
```
Comment (akien-mga) [inline editor/scene_tree_dock.cpp:1772]: While it works, I'm not sure it's a good practice per se to use a bitwise operator for booleans.

Comment (fire) [reply editor/scene_tree_dock.cpp]: Expanded to editable = editable || !root_editable;

**Comment by HeadClot** [comment]
This is really useful for me and I would love to see this pushed through. Mostly because I suffer from sensory overload and the sheer amount of nodes overwhelms me. This would be a great help for me.

**Comment by TerraDev64** [comment]
Would love this ❤

**Comment by fire** [comment]
Functionality bug. Don't merge yet, will have fix soon.

**Comment by fire** [comment]
Add test case. 
[New Game Project.zip](https://github.com/godotengine/godot/files/4833278/New.Game.Project.zip)

**Comment by groud** [comment]
I am not against the idea and understand the feature, but if it's more usable in the general case to use the "Edit all children" instead of the "Edit children" option, wouldn't it make sense to instead make the "edit children" option display all the child nodes recursively.
I don't think there's a need to setting the property on all nodes but only to the one on the top, but then the editor would instead now display the whole hierarchy.

**Comment by fire** [comment]
Aliasing "Edit children" to be using the code of "Edit all children" would be more usable. I don't mind this change, I had no feedback.

Will make the change in a bit but not right now.

**Comment by groud** [comment]
> Aliasing "Edit children" to be using the code of "Edit all children" would be more usable. I don't mind this change, I had no feedback.

This is not what I said. Only the flag would be set on the top node, exactly what is the current implementation of "Edit children" is doing.
The difference is that the flag on one node would trigger the expand of all nodes of its hierarchy, without the need to set the flag on them.

**Comment by fire** [comment]
Is this what you mean?

 https://imgur.com/a/kExKl8N

**Comment by fire** [comment]
@groud Can you repeat your comment or make a pr to my branch?

https://github.com/godotengine/godot/pull/39306#issuecomment-726152913 I don't understand.

**Comment by groud** [comment]
Sorry, I checked the gif this is not what I meant. To explain it point by point:
- Don't add an "Edit all children" option,
- Make it so that settings the flag on a scene expands all its children **recursively**. Only sub scenes would not be expanded (and thus require the flag for expansion too)

And optionally:
- As the flag is useful only for scenes now, don't display the option on nodes that are not scenes.

**Comment by fire** [comment]
This is salvagable.

**Pull request closed without merging**